src/jit/morph.cpp

```diff
@@ -6110,6 +6110,7 @@ GenTree* Compiler::fgMorphField(GenTree* tree, MorphAddrContext* mac)
     GenTree*             objRef          = tree->gtField.gtFldObj;
     bool                 fieldMayOverlap = false;
     bool                 objIsLocal      = false;
+    bool                 fieldAddr       = (tree->gtFlags & GTF_DONT_CSE) != 0;
```

Comment:
Could you name this `isFieldAddr` and add some comments? In particular, that this is intending to capture the case where this `FIELD` is simply being used in an addressing context.

Also, it seems a bit dicey to me to rely on the `GTF_DONT_CSE` flag for this. That flag generally means either "this isn't idempotent" or "this is unlikely to be profitable to CSE", and here you're relying on it to mean "This may be part of an address computation". This may be an OK workaround, but it seems to merit some pretty serious caution comments.

Reply:
I'm also worried about using `GTF_DONT_CSE` for this. 
Your comment in #23791 is
```
Looks like morph is losing track of context in ADDR/FIELD/ADDR/FIELD chains (as in the "one more ldflda case" and in the initial example). fgMorphField will modify the FIELD into an IND, and this causes morph to reset the address context as it thinks the IND will actually do a deref. But that's not true if the parent of the IND is an ADDR, as morph will later change ADDR(IND(X)) to X.
```
Can't fgMorphField take the context into account and not reset context if the current context is ADDR? I haven't thought this through completely but we can even add a new MorphAddrContextKind if necessary.

Reply:
The drawback of using GTF_DONT_CSE here is that we will sometimes explicitly null check when it's not really needed, as we continue using the current context in some cases, rather than starting a new one. So this is not an ideal fix but I don't see it as particularly risky (other than the usual risk of perturbing IR by adding commas, etc). The offset value in the context only steers the checking logic, it won't (say) rematerialize as a constant in the IR.

But maybe I'm overlooking something?

My first impulse was to handle this via the `MorphAddressContext`, but it is not fine-grained enough. In the bug case we're in and IND context when we process the field. The kind field in the context doesn't always reflect the parent node. It may be determined by an ancestor.

We could possibly add logic so that the context could capture information about the parent node too, but we'd need to be setting and resetting this continually, and that seemed like it would be hard to get right.

Reply:
I also have reservations about using the  `GTF_DONT_CSE` flag.

Reply:
I don't like this new implicit contract between the code that sets `GTF_DONT_CSE`  and this logic. If one considers changing the code that sets `GTF_DONT_CSE` in the future, it won't be obvious that it will affect null-checking logic. I understand that we already have lots of these non-obvious implicit contracts, but I would prefer not to add another one.

I don't completely understand why we are losing context. Can you describe the context changes in the bug example `ADDR/FIELD/ADDR/FIELD`?

Reply:
It does not seem that all implicit -- `fgMorphSmpOp` explicitly sets `GTF_DONT_CSE` for `GT_ADDR` children "pre order". So it will be set on all the cases we care about... and more.

https://github.com/dotnet/coreclr/blob/29fabe7114d006af73ef768c529b4477b355a284/src/jit/morph.cpp#L11344-L11348

As far as where we were losing context:
* `fgMorphField` did not pass any context when it remorphed the tree (which was now a `GT_IND`). Perhaps if the last arg was not optional that choice would have been more obvious.
* `fgMorphSmpOp` would usually start a new context for children of `GT_IND`, and would always do so when given a null context.

So the changes I made are basically: if we suspect `GT_FIELD` or `GT_IND` nodes are really parts of address computations, just keep the existing context.

Reply:
When we are processing the left `FIELD` the context argument we are getting is `MACK_Addr`, correct? Then we transform `FIELD` to `IND` and call fgMorphSmpOp on the `IND`. Shouldn't we be passing `MACK_Addr` to fgMorphSmpOp if the fgMorphField argument was `MACK_Addr`?

Reply:
> When we are processing the left FIELD the context argument we are getting is MACK_Addr, correct?  

No, ADDR will not override a MACK_Ind, see here:

https://github.com/dotnet/coreclr/blob/29fabe7114d006af73ef768c529b4477b355a284/src/jit/morph.cpp#L11878-L11894

Reply:
FWIW, the `GTF_DONT_CSE` is also used for the same purpose in `fgMorphLocalVar`. That's in general a bad idea and it was partially responsible for a bug too: #19599

But in this case it's probably fine as the worst outcome of having it set for reasons other than being address taken is an extra null check. And considering that a proper fix for this is likely far more complex, well...

Reply:
> I don't like this new implicit contract between the code that sets GTF_DONT_CSE and this logic. If one considers changing the code that sets GTF_DONT_CSE in the future, it won't be obvious that it will affect null-checking logic.

I agree, but I also agree with @mikedn that this may be the best practical fix for the moment. If we go this route, however, I would like to see us clearly indicate 1) at the declaration of `GTF_DONT_CSE`, 2) The code in morph that sets it, and 3) the places that use it for this purpose, that it must be set on nodes that appear to be indirections but are actually part of an address computation.

Reply:
@AndyAyersMS Thank you for the explanation.

Isn't it the case that fgMorphField will get a non-null context only if its parent is GT_ADDR? The lines below seem to ensure that. If that's the case we don't need to rely on GTF_DONT_CSE.

https://github.com/dotnet/coreclr/blob/311b5e2fe413c6c74a2a3680ab54d8a978651472/src/jit/morph.cpp#L11939-L11951

Reply:
I don't think so -- we could have for example IND(FIELD) and the FIELD would see a MACK_Ind, just like it would if we had IND(ADDR(FIELD)).

Reply:
In the case of IND(FIELD) when `tree` is `IND` we will set context to `MACK_Ind` here
https://github.com/dotnet/coreclr/blob/311b5e2fe413c6c74a2a3680ab54d8a978651472/src/jit/morph.cpp#L11904
but then change it to nullptr here
https://github.com/dotnet/coreclr/blob/311b5e2fe413c6c74a2a3680ab54d8a978651472/src/jit/morph.cpp#L11939-L11951
before calling `fgMorphTree` (that calls `fgMorphField`):
https://github.com/dotnet/coreclr/blob/311b5e2fe413c6c74a2a3680ab54d8a978651472/src/jit/morph.cpp#L11953

What am I missing?

Reply:
Ah, right... had forgotten about that latter bit. So we could perhaps get rid of the GTF_DONT_CSE check for FIELD and have it always pass whatever mac it gets when remorphing, but we'd still need to keep the part where we consult GTF_DONT_CSE for IND to make sure the mac does not subsequently get reset.

Reply:
We can extend the latter bit to do the same for indirections, e.g., 

```
if (((op1->gtOper == GT_FIELD) || (op1->gtOper == GT_IND)) && (tree->gtOper != GT_ADDR)) 
 { 
     subMac1 = nullptr; 
  
     // The impact of this field's value to any ongoing 
     // address computation is handled below when looking 
     // at op2. 
 } 
```
(Not sure if in addition to `GT_IND` we need to have `GT_OBJ`, `GT_BLK`, and `GT_DYN_BLK`, in which case we may use `OperIsIndir`).

Then we won't need the other change that consults `GTF_DONT_CSE`.

Reply:
I don't think this will work in this case -- when we remorph in `fgMorphField`, we remorph starting at the IND.

Reply:
I don't see why that's a problem. The parent of `GT_IND` was already through this logic before `fgMorphField` was called. So, if we pass the context unconditionally when re-morphing from `fgMorphField`, it will be nullptr if the parent of `IND` was not `GT_ADDR`. Then it will be set to `MAC_Ind` when morphing `GT_IND`.

Reply:
Right, the remorphed IND is going to reset the mac, which is wrong, since (if the parent is ADDR) the IND is part of the ongoing address computation, just like the FIELD was.

Reply:
If ADDR(FIEELD(op))) gets a non-null context, it will be passed to gtMorphField. Then gtMorphField will call fgMorphSmpOp on IND(op) and will pass the non-null context it got. If you remove the `GTF_DONT_CSE` part from your check (but keep the null check), it will become

```
                case GT_OBJ:
                case GT_BLK:	                case GT_BLK:
                case GT_DYN_BLK:	                case GT_DYN_BLK:
                case GT_IND:	                case GT_IND:
                    subMac1 = &subIndMac1;	                    // If not part of an address computation, use the new mac.
                    if (subMac1 == nullptr)
                    {
                        subMac1 = &subIndMac1;
                    }
```
so mac won't be reset.

Reply:
I updated the last comment with a correct code snippet.

Reply:
If we do that we'll potentially introduce more unnecessary checks. Some INDs should start new macs even if when a mac is passed in, eg `IND(ADD(x, IND(y))`.

Let me try it and see; maybe those cases are rare.

Reply:
Looks like they are indeed rare, no FX diffs beween  the latest fix (which has no dependence on GTF_DONT_CSE) and my original.

Reply:
I think you also need to change

```
        if ((op1->gtOper == GT_FIELD) && (tree->gtOper != GT_ADDR))
        {
            subMac1 = nullptr;

            // The impact of this field's value to any ongoing
            // address computation is handled below when looking
            // at op2.
        }
```

to
```
        if (((op1->gtOper == GT_FIELD) || op1->OperIsIndir()) && (tree->gtOper != GT_ADDR))
        {
            subMac1 = nullptr;

            // The impact of this field's value to any ongoing
            // address computation is handled below when looking
            // at op2.
        }

```
and

```
        // If gtOp2 is a GT_FIELD, we must be taking its value,
        // so it should evaluate its address in a new context.
        if (op2->gtOper == GT_FIELD)
        {
            // The impact of this field's value to any ongoing
            // address computation is handled above when looking
            // at op1.
            mac = nullptr;
        }
```
to

```
        // If gtOp2 is a GT_FIELD, we must be taking its value,
        // so it should evaluate its address in a new context.
        if ((op2->gtOper == GT_FIELD) || op2->OperIsIndir())
        {
            // The impact of this field's value to any ongoing
            // address computation is handled above when looking
            // at op1.
            mac = nullptr;
        }
```

That should take care of `IND(ADD(x, IND(y))` case. With these changes I don't see any cases where we may introduce unnecessary checks.

Reply:
Turns out we wouldn't ever null check a pure IND case anyways, as null checking requires a FIELD.

Latest commit has the extensions to the mac reset logic.